Ruby/blocl_proc_lambda.md

```diff
@@ -0,0 +1,365 @@
+# Block, Proc, Lambda
+
+## 1. Block
+
+### Block là gì ?
+Block thật ra chỉ là một đoạn code được đặt trong cặp dấu `{}` hoặc `do .. end` và k phải là đối tượng. Như ví dụ sau:
+```
+array = [1, 2, 3, 4, 5]
+array.each{|x| puts x}
+```
+Đoạn code này thực hiện việc lặp qua các phần tử qua phương thức `.each` sau đó thực hiện lệnh trong dấu ngoặc nhọn là in giá trị truyền vào qua mỗi lần lặp ra ngoài màn hình.
+Đơn giản đúng không!. Vậy chúng ta cùng đặt ra câu hỏi là block sinh ra để làm gì nhỉ ?
+### Block để làm gì ?
+- Sử dụng cho các phương thức ở dạng tổng quát như `each` hoặc `map` hay nói chung là nhóm Iterator. Tổng quát ở đây nói tới phương thức cung cấp cho ta cách truy cập dữ liệu, trả về cho ta cái gì nhưng không lộ ra cách nó thực hiện như thế nào. Đi vào thằng `each`, khi sử dụng thì nó sẽ lặp qua các phần tử trong mảng, việc mỗi phần tử sau khi lấy ra xử lý như thế nào thì nó đếch quan tâm. Mà đương nhiên nó sẽ không trả cho ta cái gì cả. Bây giờ, nếu muốn in ra màn hình mỗi phần tử thì làm như thế nào ? Ta sẽ dùng block, và trong block ta sẽ xử lý từng phần tử đó theo ý ta.
+- Chúng ta có thể DRY code của mình. 
+### Sử dụng nó như thế nào ?
+Dù mình có đề cập tới `each` hay Iterator nhưng không hoàn toàn block chỉ sài cho mấy cái vòng lặp thôi nhé. Hãy cùng đi qua một vài ví dụ về sử dụng block nào ?
```

Comment:
xài nhé :D

Reply:
"không hoàn toàn block chỉ sài cho mấy cái vòng lặp thôi nhé"
chỗ này hơi chuối ạ